airflow/providers/teradata/transfers/teradata_to_teradata.py

```diff
@@ -0,0 +1,91 @@
+# under the License.
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Sequence
+
+from airflow.models import BaseOperator
+from airflow.providers.teradata.hooks.teradata import TeradataHook
+
+if TYPE_CHECKING:
+    from airflow.utils.context import Context
+
+
+class TeradataToTeradataOperator(BaseOperator):
+    """
+    Moves data from Teradata source database to Teradata destination database.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:TeradataToTeradataOperator`
+
+    :param teradata_destination_conn_id: destination Teradata connection.
+    :param destination_table: destination table to insert rows.
+    :param teradata_source_conn_id: :ref:`Source Teradata connection <howto/connection:Teradata>`.
+    :param source_sql: SQL query to execute against the source Teradata database
+    :param source_sql_params: Parameters to use in sql query.
+    :param rows_chunk: number of rows per chunk to commit.
+    """
+
+    template_fields: Sequence[str] = (
+        "source_sql",
+        "source_sql_params",
+    )
+    template_ext: Sequence[str] = (".sql",)
+    template_fields_renderers = {"source_sql": "sql", "source_sql_params": "py"}
+    ui_color = "#e07c24"
+
+    def __init__(
+        self,
+        *,
+        teradata_destination_conn_id: str,
+        destination_table: str,
+        teradata_source_conn_id: str,
+        source_sql: str,
+        source_sql_params: dict | None = None,
+        rows_chunk: int = 5000,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        if source_sql_params is None:
+            source_sql_params = {}
+        self.teradata_destination_conn_id = teradata_destination_conn_id
+        self.destination_table = destination_table
+        self.teradata_source_conn_id = teradata_source_conn_id
+        self.source_sql = source_sql
+        self.source_sql_params = source_sql_params
+        self.rows_chunk = rows_chunk
+
+    def _execute(self, src_hook, dest_hook, context) -> None:
+        with src_hook.get_conn() as src_conn:
+            cursor = src_conn.cursor()
+            cursor.execute(self.source_sql, self.source_sql_params)
+            target_fields = [field[0] for field in cursor.description]
+            rows_total = 0
+            for rows in iter(lambda: cursor.fetchmany(self.rows_chunk), []):
+                dest_hook.bulk_insert_rows(
+                    self.destination_table, rows, target_fields=target_fields, commit_every=self.rows_chunk
+                )
+                rows_total += len(rows)
+            self.log.info("Finished data transfer.")
+            cursor.close()
+
+    def execute(self, context: Context) -> None:
+        src_hook = TeradataHook(teradata_conn_id=self.teradata_source_conn_id)
+        dest_hook = TeradataHook(teradata_conn_id=self.teradata_destination_conn_id)
+        self._execute(src_hook, dest_hook, context)
```

Comment:
Why do we need this private function?
You can mark specific helpers as private but marking the whole execute function as private is not something that we do. If allowed it means that all operators logic is considered internal detail... that doesn't make sense to me.

Reply:
I don't think this is right. It may lead to missing breaking changes by mistake.
I will add todo on the oracle to fix it as well

Reply:
Okay. We added this to our backlong. Will work on it in next release.

Reply:
@Airflow Team, if Elad is caught up and can't look into this, is there someone else who could review? We'd really appreciate it!

Reply:
> However, we're facing challenges in executing the unit tests (after changes).

What kind of the challenges?

Reply:
You could move hooks into the properties and test it separately: initialise hook, operator execution, some other cases

```python
from functools import cached_property

from airflow.models import BaseOperator


class FakeHook:
    def __init__(self, conn_id: str):
        self.conn_id = conn_id

    def method_1(self, param1, param2):
        # Should be covered into the Hook tests
        return "spam"

    def method_2(self, param1, param2):
        # Should be covered into the Hook tests
        return "egg"

class SampleTransferOperator(BaseOperator):
    def __init__(self, *, src_conn_id: str, dest_conn_id: str, **kwargs):
        super().__init__(**kwargs)
        self.src_conn_id = src_conn_id
        self.dest_conn_id = dest_conn_id

    @cached_property
    def src_hook(self) -> FakeHook:
        return FakeHook(conn_id=self.src_conn_id)

    @cached_property
    def dest_hook(self) -> FakeHook:
        return FakeHook(conn_id=self.dest_conn_id)

    def execute(self, context):
        return f"{self.src_hook.method_1(1, 2)} {self.dest_hook.method_2(3, 4)}"


#### Test Cases for operator
from unittest import mock
import pytest

@pytest.fixture
def mocked_src_hook():
    with mock.patch.object(SampleTransferOperator, "src_hook", spec=FakeHook, name="FakeSourceHook") as m:
        yield m

@pytest.fixture
def mocked_dest_hook():
    with mock.patch.object(SampleTransferOperator, "dest_hook", spec=FakeHook, name="FakeDestHook") as m:
        yield m


class TestSampleTransferOperator:
    def test_source_hook(self):
        op = SampleTransferOperator(task_id="test_source_hook", src_conn_id="foo", dest_conn_id="bar")
        hook = op.src_hook
        # If the assumption that hook already tested correct then it would be enough to test assignments
        assert hook
        assert hook is op.src_hook
        assert hook.conn_id == "foo"

    def test_destination_hook(self):
        op = SampleTransferOperator(task_id="test_source_hook", src_conn_id="foo", dest_conn_id="bar")
        hook = op.dest_hook
        # If the assumption that hook already tested correct then it would be enough to test assignments
        assert hook
        assert hook is op.dest_hook
        assert hook.conn_id == "bar"

    def test_execution(self, mocked_src_hook, mocked_dest_hook):
        mocked_src_hook.method_1.return_value = "baz"
        mocked_dest_hook.method_2.return_value = "qux"

        op = SampleTransferOperator(task_id="test_execution", src_conn_id="foo", dest_conn_id="bar")
        result = op.execute({})
        assert result == "baz qux"
        mocked_src_hook.method_1.assert_called_once_with(1, 2)
        mocked_dest_hook.method_2.assert_called_once_with(3, 4)
```

Reply:
Yeah maybe just push the change removing `_execute` method so we can help debug.

I don't really see a major problem with it strictly speaking.  It's unnecessary obfuscation to add the _execute method it seems, but it also isn't against any rules or conventions i'm aware of.... but yeah we can help debug

Reply:
Modified teradata_to_teradata.py by removing _execute method and corresponding unit test is test_teradata_to_teradata.py modified to reflect operator change. Tests are working fine.

Reply:
> You could move hooks into the properties and test it separately: initialise hook, operator execution, some other cases
> 
> ```python
> from functools import cached_property
> 
> from airflow.models import BaseOperator
> 
> 
> class FakeHook:
>     def __init__(self, conn_id: str):
>         self.conn_id = conn_id
> 
>     def method_1(self, param1, param2):
>         # Should be covered into the Hook tests
>         return "spam"
> 
>     def method_2(self, param1, param2):
>         # Should be covered into the Hook tests
>         return "egg"
> 
> class SampleTransferOperator(BaseOperator):
>     def __init__(self, *, src_conn_id: str, dest_conn_id: str, **kwargs):
>         super().__init__(**kwargs)
>         self.src_conn_id = src_conn_id
>         self.dest_conn_id = dest_conn_id
> 
>     @cached_property
>     def src_hook(self) -> FakeHook:
>         return FakeHook(conn_id=self.src_conn_id)
> 
>     @cached_property
>     def dest_hook(self) -> FakeHook:
>         return FakeHook(conn_id=self.dest_conn_id)
> 
>     def execute(self, context):
>         return f"{self.src_hook.method_1(1, 2)} {self.dest_hook.method_2(3, 4)}"
> 
> 
> #### Test Cases for operator
> from unittest import mock
> import pytest
> 
> @pytest.fixture
> def mocked_src_hook():
>     with mock.patch.object(SampleTransferOperator, "src_hook", spec=FakeHook, name="FakeSourceHook") as m:
>         yield m
> 
> @pytest.fixture
> def mocked_dest_hook():
>     with mock.patch.object(SampleTransferOperator, "dest_hook", spec=FakeHook, name="FakeDestHook") as m:
>         yield m
> 
> 
> class TestSampleTransferOperator:
>     def test_source_hook(self):
>         op = SampleTransferOperator(task_id="test_source_hook", src_conn_id="foo", dest_conn_id="bar")
>         hook = op.src_hook
>         # If the assumption that hook already tested correct then it would be enough to test assignments
>         assert hook
>         assert hook is op.src_hook
>         assert hook.conn_id == "foo"
> 
>     def test_destination_hook(self):
>         op = SampleTransferOperator(task_id="test_source_hook", src_conn_id="foo", dest_conn_id="bar")
>         hook = op.dest_hook
>         # If the assumption that hook already tested correct then it would be enough to test assignments
>         assert hook
>         assert hook is op.dest_hook
>         assert hook.conn_id == "bar"
> 
>     def test_execution(self, mocked_src_hook, mocked_dest_hook):
>         mocked_src_hook.method_1.return_value = "baz"
>         mocked_dest_hook.method_2.return_value = "qux"
> 
>         op = SampleTransferOperator(task_id="test_execution", src_conn_id="foo", dest_conn_id="bar")
>         result = op.execute({})
>         assert result == "baz qux"
>         mocked_src_hook.method_1.assert_called_once_with(1, 2)
>         mocked_dest_hook.method_2.assert_called_once_with(3, 4)
> ```

Thank you for the reference code. It helps me to resolve the issue.